File: benchmarking/benchmark.py
import subprocess
import pandas as pd
import os

import benchmarking.slides as sld


def generate_stream_runs(max_streams):
    base = [x for x in [1, 3, 5] if x <= max_streams]
    if max_streams > 5:
        base += list(range(10, max_streams + 1, 5))
    return base


def run_benchmark(
    input_file,
    streams,
    project_absolute_path,
    results_absolute_path,
    exe,
    is_single_threaded,
    is_verbose,
    write_to_csv,
):
    print(f"Running benchmark with {streams} streams...")
    result = subprocess.run(
        [
            exe,
            input_file,
            str(streams),
            results_absolute_path,
            project_absolute_path,
            str(is_single_threaded),
            str(is_verbose),
            str(write_to_csv),
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        encoding="utf-8",
    )
    if result.returncode != 0:
        print(f"Error running benchmark: {result.stderr}")
        return pd.DataFrame(), result.stdout
    return parse_output(result.stdout, streams), result.stdout


def parse_output(output_text, stream_count):
    in_table = False
    results = []
    for line in output_text.split("\n"):
        line = line.strip()
        if "Method" in line and "ms/frame(strm)" in line:
            in_table = True
            continue

        if not in_table:
            continue
        # Skip separator lines and blanks
        if (
            not line
            or line.startswith("—")
            or line.startswith("---")
            or line.startswith("===")
        ):
            continue

        if line.startswith("ms/") or line.startswith("CPU") or line.startswith("Total"):
            continue

        parts = [p.strip() for p in line.split("|")]

        # Format: 9 columns
        # [0] Method
        # [1] ms/frame(strm)   e.g. "10.23 ms"
        # [2] Total FPS        e.g. "97.8"
        # [3] CPU ms/frame     e.g. "7.321 ms"
        # [4] CPU %/stream     e.g. "85%"
        # [5] CPU % total      e.g. "87%"
        # [6] Mem Total KB
        # [7] Mem/Strm KB
        # [8] Frames
        if len(parts) < 9:
            continue
        try:
            method = parts[0]
            time_per_frame = float(parts[1].replace("ms", "").strip())
            fps = float(parts[2].strip())
            cpu_ms_per_frame = float(parts[3].replace("ms", "").strip())
            mem_total_kb = float(parts[6].strip())
            mem_per_stream_kb = float(parts[7].strip())
            frames = int(parts[8].strip())

            results.append(
                {
                    "method": method,
                    "streams": stream_count,
                    "time_per_frame": time_per_frame,
                    "fps": fps,
                    "cpu_ms_per_frame": cpu_ms_per_frame,
                    "memory": mem_total_kb,
                    "mem_per_stream": mem_per_stream_kb,
                    "frames": frames,
                }
            )
        except (ValueError, IndexError):
            pass

    return pd.DataFrame(results)


def run_benchmark_averaged(
    input_file,
    streams,
    project_absolute_path,
    results_absolute_path,
    exe,
    is_single_threaded,
    is_verbose,
    write_to_csv,
    n_runs,
):
    numeric_cols = [
        "time_per_frame",
        "fps",
        "cpu_ms_per_frame",
        "memory",
        "mem_per_stream",
        "frames",
    ]

    run_dfs = []
    for run_idx in range(1, n_runs + 1):
        print(f"  Run {run_idx}/{n_runs} for {streams} streams...")
        df, _ = run_benchmark(
            input_file,
            streams,
            project_absolute_path,
            results_absolute_path,
            exe,
            is_single_threaded,
            is_verbose,
            write_to_csv,
        )
        if not df.empty:
            run_dfs.append(df)

    if not run_dfs:
        return pd.DataFrame()

    combined = pd.concat(run_dfs, ignore_index=True)
    averaged = combined.groupby(["method", "streams"], as_index=False)[
        numeric_cols
    ].mean()
    return averaged


def benchmark(
    input,
    streams,
    executable_absolute_path,
    exe,
    project_absolute_path,
    results_absolute_path,
    slides_config,
    plots_folder,
    is_single_threaded,
    is_verbose,
    write_to_csv,
    video_type,
    n_runs=1,
):
    exe_fullpath = os.path.join(executable_absolute_path, exe)

    stream_steps = generate_stream_runs(streams)
    print(f"Stream ranges to test: {stream_steps}")
    print(f"Each stream count will be run {n_runs} time(s) and averaged.")

    all_results = []
    for s in stream_steps:
        df = run_benchmark_averaged(
            input,
            s,
            project_absolute_path,
            results_absolute_path,
            exe_fullpath,
            is_single_threaded,
            is_verbose,
            write_to_csv,
            n_runs,
        )
        if df.empty:
            print(f"Warning: No data returned for streams={s}")
        all_results.append(df)

    # Drop any empty DataFrames before concat to avoid dtype warnings
    non_empty = [df for df in all_results if not df.empty]
    if not non_empty:
        print(
            "Error: all stream runs returned empty results. Check C++ binary and output format."
        )
        return

    full_df = pd.concat(non_empty, ignore_index=True)

    csv_path = os.path.join(plots_folder, "benchmark_results.csv")
    full_df.to_csv(csv_path, index=False)
    print(f"Saved complete data table: {csv_path}")

    sld.produce_slides(
        full_df,
        slides_config,
        "benchmark_comparison_slides.pptx",
        plots_folder,
        video_type,
    )

File: benchmarking/benchmark.rs
use std::collections::HashMap;
use std::fmt::Write as FmtWrite;
use std::fs;
use std::process::Command;

use crate::benchmark_extractors::run_benchmark_extractors;

// ── Data types shared across modules ────────────────────────────────────────

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct BenchmarkResult {
    pub method: String,
    pub streams: i32,
    pub time_per_frame: f64,
    pub fps: f64,
    pub cpu_ms_per_frame: f64,
    pub memory: f64,
    pub mem_per_stream: f64,
    pub frames: i32,
}

pub const NUMERIC_FIELDS: &[&str] = &[
    "time_per_frame",
    "fps",
    "cpu_ms_per_frame",
    "memory",
    "mem_per_stream",
    "frames",
];

// ── Stream run generation ───────────────────────────────────────────────────

pub fn generate_stream_runs(max_streams: i32) -> Vec<i32> {
    let mut base: Vec<i32> = [1, 3, 5].iter().copied().filter(|&x| x <= max_streams).collect();
    if max_streams > 5 {
        let mut s = 10;
        while s <= max_streams {
            base.push(s);
            s += 5;
        }
    }
    base
}

// ── Output parsing ──────────────────────────────────────────────────────────

pub fn parse_output(output_text: &str, stream_count: i32) -> Vec<BenchmarkResult> {
    let mut in_table = false;
    let mut results = Vec::new();

    for line in output_text.lines() {
        let line = line.trim();

        if line.contains("Method") && line.contains("ms/frame(strm)") {
            in_table = true;
            continue;
        }
        if !in_table {
            continue;
        }
        if line.is_empty()
            || line.starts_with('\u{2014}')
            || line.starts_with("---")
            || line.starts_with("===")
        {
            continue;
        }
        if line.starts_with("ms/") || line.starts_with("CPU") || line.starts_with("Total") {
            continue;
        }

        let parts: Vec<&str> = line.split('|').map(|p| p.trim()).collect();
        if parts.len() < 9 {
            continue;
        }

        let parse_f = |s: &str| -> Option<f64> {
            s.replace("ms", "").trim().parse::<f64>().ok()
        };

        let method = parts[0].to_string();
        let time_per_frame = match parse_f(parts[1]) { Some(v) => v, None => continue };
        let fps = match parts[2].parse::<f64>() { Ok(v) => v, Err(_) => continue };
        let cpu_ms_per_frame = match parse_f(parts[3]) { Some(v) => v, None => continue };
        let mem_total_kb = match parts[6].parse::<f64>() { Ok(v) => v, Err(_) => continue };
        let mem_per_stream_kb = match parts[7].parse::<f64>() { Ok(v) => v, Err(_) => continue };
        let frames = match parts[8].parse::<i32>() { Ok(v) => v, Err(_) => continue };

        results.push(BenchmarkResult {
            method,
            streams: stream_count,
            time_per_frame,
            fps,
            cpu_ms_per_frame,
            memory: mem_total_kb,
            mem_per_stream: mem_per_stream_kb,
            frames,
        });
    }
    results
}

// ── Run benchmark subprocess ────────────────────────────────────────────────

pub fn run_benchmark(
    input_file: &str,
    streams: i32,
    project_absolute_path: &str,
    results_absolute_path: &str,
    is_single_threaded: i32,
    is_verbose: i32,
    write_to_csv: i32,
) -> Vec<BenchmarkResult> {
    println!("Running benchmark with {} streams...", streams);

    run_benchmark_extractors(
        input_file,
        streams,
        results_absolute_path,
        project_absolute_path,
        is_single_threaded != 0,
        is_verbose != 0,
        write_to_csv != 0,
    )
    .unwrap_or_default()
}

// ── Averaged runs ───────────────────────────────────────────────────────────

pub fn run_benchmark_averaged(
    input_file: &str,
    streams: i32,
    project_absolute_path: &str,
    results_absolute_path: &str,
    is_single_threaded: i32,
    is_verbose: i32,
    write_to_csv: i32,
    n_runs: usize,
) -> Vec<BenchmarkResult> {
    let mut all_runs: Vec<Vec<BenchmarkResult>> = Vec::new();

    for run_idx in 1..=n_runs {
        println!("  Run {}/{} for {} streams...", run_idx, n_runs, streams);
        let results = run_benchmark(
            input_file,
            streams,
            project_absolute_path,
            results_absolute_path,
            is_single_threaded,
            is_verbose,
            write_to_csv,
        );
        if !results.is_empty() {
            all_runs.push(results);
        }
    }

    if all_runs.is_empty() {
        return Vec::new();
    }

    let mut groups: HashMap<(String, i32), Vec<&BenchmarkResult>> = HashMap::new();
    for run in &all_runs {
        for r in run {
            groups
                .entry((r.method.clone(), r.streams))
                .or_default()
                .push(r);
        }
    }

    let mut averaged = Vec::new();
    for ((method, streams_val), entries) in &groups {
        let n = entries.len() as f64;
        averaged.push(BenchmarkResult {
            method: method.clone(),
            streams: *streams_val,
            time_per_frame: entries.iter().map(|e| e.time_per_frame).sum::<f64>() / n,
            fps: entries.iter().map(|e| e.fps).sum::<f64>() / n,
            cpu_ms_per_frame: entries.iter().map(|e| e.cpu_ms_per_frame).sum::<f64>() / n,
            memory: entries.iter().map(|e| e.memory).sum::<f64>() / n,
            mem_per_stream: entries.iter().map(|e| e.mem_per_stream).sum::<f64>() / n,
            frames: (entries.iter().map(|e| e.frames as f64).sum::<f64>() / n) as i32,
        });
    }
    averaged
}

// ── CSV writing ─────────────────────────────────────────────────────────────

pub fn write_csv(results: &[BenchmarkResult], path: &str) -> Result<(), String> {
    let mut buf = String::new();
    writeln!(
        buf,
        "method,streams,time_per_frame,fps,cpu_ms_per_frame,memory,mem_per_stream,frames"
    )
    .unwrap();
    for r in results {
        writeln!(
            buf,
            "{},{},{},{},{},{},{},{}",
            r.method, r.streams, r.time_per_frame, r.fps, r.cpu_ms_per_frame, r.memory,
            r.mem_per_stream, r.frames
        )
        .unwrap();
    }
    fs::write(path, buf).map_err(|e| format!("Failed to write CSV: {}", e))
}

// ── Main benchmark orchestrator ─────────────────────────────────────────────

pub fn benchmark(
    input: &str,
    streams: i32,
    project_absolute_path: &str,
    results_absolute_path: &str,
    slides_config: &str,
    plots_folder: &str,
    is_single_threaded: i32,
    is_verbose: i32,
    write_to_csv: i32,
    video_type: &str,
    n_runs: usize,
) {
    let stream_steps = generate_stream_runs(streams);
    println!("Stream ranges to test: {:?}", stream_steps);
    println!(
        "Each stream count will be run {} time(s) and averaged.",
        n_runs
    );

    let mut all_results: Vec<BenchmarkResult> = Vec::new();

    for &s in &stream_steps {
        let results = run_benchmark_averaged(
            input,
            s,
            project_absolute_path,
            results_absolute_path,
            is_single_threaded,
            is_verbose,
            write_to_csv,
            n_runs,
        );
        if results.is_empty() {
            println!("Warning: No data returned for streams={}", s);
        }
        all_results.extend(results);
    }

    if all_results.is_empty() {
        println!("Error: all stream runs returned empty results. Check C++ binary and output format.");
        return;
    }

    let csv_path = format!("{}/benchmark_results.csv", plots_folder);
    if let Err(e) = write_csv(&all_results, &csv_path) {
        eprintln!("{}", e);
    } else {
        println!("Saved complete data table: {}", csv_path);
    }

    // ── Rust slide generation (disabled for now) ──
    // slides::produce_slides(
    //     &all_results,
    //     slides_config,
    //     "benchmark_comparison_slides.pptx",
    //     plots_folder,
    //     video_type,
    // );

    // ── Python slide generation via benchmarking/slides.py ──
    println!("Running Python slide generation via benchmarking.slides...");
    let venv_python = format!("{}/../venv-motion-vectors/bin/python3", project_absolute_path);
    let py_code = format!(
        "import pandas as pd; \
         import benchmarking.slides as s; \
         df = pd.read_csv('{csv}'); \
         s.produce_slides(df, '{cfg}', 'benchmark_comparison_slides.pptx', '{plots}', '{vtype}')",
        csv = csv_path,
        cfg = slides_config,
        plots = plots_folder,
        vtype = video_type,
    );

    let status = Command::new(&venv_python)
        .args(["-c", &py_code])
        .current_dir(project_absolute_path)
        .status();

    match status {
        Ok(s) if s.success() => println!("Python slide generation complete."),
        Ok(s) => eprintln!("Python slide generation exited with status: {}", s),
        Err(e) => eprintln!("Failed to run python slide generation: {}", e),
    }
}
